src/dag/write.ts

```diff
@@ -0,0 +1,158 @@
+import {getRootAsMeta} from './meta.js';
+import type * as kv from '../kv/mod.js';
+import {HeadChange, toLittleEndian, fromLittleEndian} from './dag.js';
+import {chunkDataKey, chunkMetaKey, headKey, chunkRefCountKey} from './key.js';
+import {Read} from './read.js';
+import type {Chunk} from './chunk.js';
+import * as utf8 from '../utf8.js';
+
+export class Write {
+  private readonly _kvw: kv.Write;
+  private readonly _mutatedChunks = new Set<string>();
+  private readonly _changedHeads = new Map<string, HeadChange>();
+
+  constructor(kvw: kv.Write) {
+    this._kvw = kvw;
+  }
+
+  read(): Read {
+    return new Read(this._kvw);
+  }
+
+  async putChunk(c: Chunk): Promise<void> {
+    const {hash, data, meta} = c;
+    const key = chunkDataKey(hash);
+    const p1 = this._kvw.put(key, data);
```

Comment:
We need to be clear about what the semantics are of inserting a chunk that already exists.

IIRC, the desired contract is that it is invalid to reuse a hash but with different data or meta, and that the behavior of doing that is simply undefined.

If this is correct then we should have some debug-only code ensuring that invariant.